### PR TITLE
refactor(url): layered grammar-driven URL parser with precise diagnostics (#477)

### LINKED ISSUE #477: Refactor URL parser to follow the formal ABNF spec
# Refactor URL parser to follow the formal ABNF spec (§6 layered architecture)

> **Note:** This issue was written by Claude Sonnet 4.6 (AI) at the request of the project maintainer.

---

## Background

The project has a detailed formal URL contract spec in
[`docs/04-feature-specs/shareable-urls-formal.md`](docs/04-feature-specs/shareable-urls-formal.md)
(Final v6, April 2026). It defines:

- An **ABNF grammar** for all query-string parameters (§2)
- A **semantic resolution pipeline** (§3.1 — 14 numbered steps)
- A **canonicalization algorithm** with strict parameter ordering (§4)
- **21 conformance test vectors** covering valid inputs, recovery, and edge cases (§5)
- A **recommended implementation architecture** (§6)

The spec's §6 recommends four clean, layered functions:

```
parseQuery(raw: string): ParsedTokens          // ABNF-equivalent tokenizer
resolveState(tokens, route, services): ResolvedState  // semantic pass
canonicalize(state): string                    // single canonical URL writer
migrateUrl(vFrom, vTo, state): state           // version migration chain
```

---

## Problem: current code does not follow the spec architecture

The current implementation collapses all four spec phases into two monolithic
decode functions that mix tokenization, semantic resolution, validation,
defaulting, and fallback logic in a single pass:

- [`src/lib/utils/calculator-url.ts` — `decodeCalculatorUrl()`](src/lib/utils/calculator-url.ts#L450)
- [`src/lib/utils/plot-url.ts` — `decodePlotUrl()`](src/lib/utils/plot-url.ts#L230)

### Specific deviations from the spec

| Spec requirement | Current state |
|---|---|
| `parseQuery` — pure ABNF tokenization, no semantics | Absent. Uses `URLSearchParams.get()` + ad-hoc `split(",")` / `lastIndexOf(":")` directly inside the decode functions |
| `resolveState` — separate semantic pass (mode gating, defaults, validation, precedence) | Merged inline into `decodeCalculatorUrl` / `decodePlotUrl` |
| `canonicalize` — one canonical URL writer shared across calculator and plot | Split into `encodeCalculatorUrl` + `calculatorUrlQueryString` (calculator) and `encodePlotUrl` + `plotUrlQueryString` (plot) with no shared `ResolvedState` type; correctness of canonical ordering is untested structurally |
| `migrateUrl` — version migration chain | [`_urlv` is parsed and immediately discarded with `void _urlv`](src/lib/utils/calculator-url.ts#L475-L476) — no migration logic exists |
| Spec §3.2: duplicate resolution is a distinct step | Implemented as `resolveLastWins()` helper but called inline, not as an explicit pipeline stage |
| Spec §3.1 step 5: `extdata` is extracted before duplicate resolution | Handled correctly (a workaround comment notes why), but the necessity is a symptom of the merged pipeline |
| Spec §3.5: advanced-mode gating must be enforced as a single conditional-enablement pass | Scattered across many `if (isAdvancedMode)` blocks inside the single decode function |

### Concrete impact

1. **Untestable in isolation.** There is no way to unit-test tokenization
   without also exercising semantic resolution and defaulting. Any test
   must provide a fully valid `URLSearchParams` object.

2. **No migration path.** The spec defines a `migrateUrl` chain for future
   major version bumps (§3.4). The current code has a comment placeholder
   but no actual chain. Adding migration in the current architecture means
   adding more branches to an already large function.

3. **Canonicalization is not a pure function of state.** The encoder takes
   `CalculatorUrlState` / `PlotUrlInput` (which includes raw inputs), not a
   `ResolvedState`. This makes it harder to guarantee that the same logical
   state always produces the same URL.

4. **Drift risk between calculator and plot decoders.** Duplicate-Z
   collapsing logic, `parseStrictFiniteNumber`, `parseStrictAtomicNumber`,
   `resolveLastWins`, and the entire `mat_*` block are copy-pasted between
   the two files. A future change to one is likely to miss the other.

---

### 1. Introduce a `parseQuery` stage

Extract all ABNF tokenization into a pure function:

```ts
// New file: src/lib/utils/url-parse.ts
function parseQuery(raw: string | URLSearchParams): ParsedTokens
```

`ParsedTokens` is a plain record of raw string values for each known
parameter key (after duplicate resolution and `extdata` extraction), plus
an ordered list of `extdata` sources. No semantics, no defaulting, no
type coercion.

This function should be trivially unit-testable against the 21 conformance
vectors in §5 of the formal spec.

### 2. Introduce a `resolveState` stage

All mode gating (§3.5), defaults (§3.6), validation (§3.7), and conditional
enablement currently scattered through `decodeCalculatorUrl` /
`decodePlotUrl` should become a single:

```ts
function resolveCalculatorState(tokens: ParsedTokens): ResolvedCalculatorState
function resolvePlotState(tokens: ParsedTokens): ResolvedPlotState
```

`ResolvedState` types carry fully typed, validated, defaulted values — no
raw strings, no `undefined` where a default applies.

### 3. Implement the `migrateUrl` chain

```ts
function migrateUrl(vFrom: number, vTo: number, tokens: ParsedTokens): ParsedTokens
```

Even though `CURRENT_URL_MAJOR = 1` currently and there is nothing to
migrate, the chain must exist so that a future major bump can be wired in
without touching `resolveState`. The version-mismatch banner
([`src/lib/components/url-version-warning-banner.svelte`](src/lib/components/url-version-warning-banner.svelte))
and the version logic
([`src/lib/utils/url-version.ts`](src/lib/utils/url-version.ts))
should hook into this chain.

### 4. Consolidate shared helpers

Move `resolveLastWins`, `parseStrictFiniteNumber`, `parseStrictAtomicNumber`,
and the `mat_*` element parsing block to a single internal module so they
are not duplicated between `calculator-url.ts` and `plot-url.ts`.

### 5. Keep existing public API stable (or update call sites)

The public signatures `decodeCalculatorUrl` / `decodeCalculatorUrl` /
`encodePlotUrl` / `calculatorUrlQueryString` / `plotUrlQueryString` are
called from:

- [`src/routes/calculator/+page.svelte`](src/routes/calculator/+page.svelte)
- [`src/routes/plot/+page.svelte`](src/routes/plot/+page.svelte)
- Various state modules under [`src/lib/state/`](src/lib/state/)

These can remain as thin wrappers over the new layered internals, or call
sites can be updated — either approach is acceptable as long as the
observable URL behavior is unchanged.

---

## Acceptance criteria

- [ ] A `parseQuery` function exists that accepts a raw query string (or
      `URLSearchParams`) and returns `ParsedTokens` with no semantic
      logic.
- [ ] All 21 conformance test vectors from
      [`docs/04-feature-specs/shareable-urls-formal.md` §5](docs/04-feature-specs/shareable-urls-formal.md)
      are covered by unit tests against `parseQuery`.
- [ ] A `resolveCalculatorState` and `resolvePlotState` function exist,
      separately testable without a full page render.
- [ ] A `migrateUrl` stub exists and is wired into the version-check path
      in [`src/lib/utils/url-version.ts`](src/lib/utils/url-version.ts).
- [ ] No logic is duplicated between the calculator and plot URL modules.
- [ ] All existing tests continue to pass:
  - Unit: [`src/tests/unit/calculator-url.test.ts`](src/tests/unit/calculator-url.test.ts),
    [`src/tests/unit/plot-url.test.ts`](src/tests/unit/plot-url.test.ts)
  - Contract: [`src/tests/contracts/url-codec.contract.test.ts`](src/tests/contracts/url-codec.contract.test.ts)
  - E2E: [`tests/e2e/url-parser.spec.ts`](tests/e2e/url-parser.spec.ts),
    [`tests/e2e/calculator-url.spec.ts`](tests/e2e/calculator-url.spec.ts)
- [ ] New unit tests are added to cover the `parseQuery` tokenization
      layer in isolation (conformance vectors §5.1 and §5.2).
- [ ] The E2E tests in [`tests/e2e/url-parser.spec.ts`](tests/e2e/url-parser.spec.ts)
      continue to pass against the refactored implementation (all 6 acceptance
      scenarios: urlv mismatch, load-defaults, current version, missing urlv,
      custom compound round-trip, duplicate params, unknown params dropped).
- [ ] An AI session log is written to `docs/ai-logs/` and a row is
      prepended to `CHANGELOG-AI.md` per the mandatory logging rules in
      `CLAUDE.md`.

---

## Files of interest for the implementor

| File | Role |
|---|---|
| [`docs/04-feature-specs/shareable-urls-formal.md`](docs/04-feature-specs/shareable-urls-formal.md) | **Normative spec** — ABNF grammar, semantic rules, canonicalization algorithm, conformance vectors |
| [`docs/04-feature-specs/shareable-urls.md`](docs/04-feature-specs/shareable-urls.md) | UX/product companion spec (wins on UX behavior if conflict) |
| [`src/lib/utils/calculator-url.ts`](src/lib/utils/calculator-url.ts) | Current monolithic calculator decoder/encoder |
| [`src/lib/utils/plot-url.ts`](src/lib/utils/plot-url.ts) | Current monolithic plot decoder/encoder |
| [`src/lib/external-data/url.ts`](src/lib/external-data/url.ts) | `extdata` param parsing (already reasonably isolated) |
| [`src/lib/utils/url-version.ts`](src/lib/utils/url-version.ts) | Version negotiation helpers |
| [`src/lib/utils/energy-parser.ts`](src/lib/utils/energy-parser.ts) | Energy token parser (called by encoder; keep as-is) |
| [`src/tests/contracts/url-codec.contract.test.ts`](src/tests/contracts/url-codec.contract.test.ts) | Contract tests ensuring every union member round-trips |
| [`tests/e2e/url-parser.spec.ts`](tests/e2e/url-parser.spec.ts) | Playwright E2E tests for the 6 acceptance scenarios |
| [`tests/e2e/calculator-url.spec.ts`](tests/e2e/calculator-url.spec.ts) | Broader Playwright E2E coverage for calculator URL state |

---

## Out of scope for this issue

- Changing any URL parameter names or semantics (the public URL contract is frozen at v1)
- Implementing a real `migrateUrl` migration path (only the stub + wiring is needed)
- Changing the UI or component tree
- Touching the WASM wrapper or libdedx integration

### PR DESCRIPTION
**Pull request by grzanka**
## Refactor URL parser to the §6 layered architecture (closes #477)

Replaces the two monolithic decode functions with the separately-testable stages from `docs/04-feature-specs/shareable-urls-formal.md` §6, **plus** span-accurate error messages for malformed URLs (per maintainer request on the issue thread).

> **Scope (agreed on the issue thread):** v2-only; v1 links are rejected with an "unsupported link" banner, not migrated. Observable URL behaviour is otherwise **preserved** — the existing ~1600-test suite is the oracle. Full design: https://github.com/APTG/dedx_web/issues/477#issuecomment-4572154960

### Architecture (implemented)

```
parseQuery(raw)             -> QueryNode (AST)   — Peggy grammar, pure syntax, source spans
resolveCalculatorState(ast) -> CalculatorUrlState + Diagnostic[]
resolvePlotState(ast)       -> PlotUrlDecoded    + Diagnostic[]
decode*()                   = parseQuery -> resolve*  (public API unchanged)
migrateUrl(from,to,tokens)  — identity seam in url-version.ts (v1 rejected, not migrated)
```

- **Grammar as code:** `url-grammar.peggy` compiled by an inline Vite loader (shared by Vitest; no committed generated parser). Structured AST for list params with a graceful `raw` fallback so malformed values are handled exactly as before.
- **Diagnostics:** one `Diagnostic { severity, code, message, span, param }`. Syntax failures throw `UrlParseError` with the exact offset; `renderCaret` underlines the offending text; named rules give friendly "Expected … but found …". Unknown/invalid params produce info/warning diagnostics carrying spans.
- **De-duplication:** `resolveLastWins`, `parseStrict*`, the `mat_*` block, and `mat_elements` encoding now live once in `url-shared.ts`.

### Status — all acceptance criteria met

- [x] `parseQuery` exists, pure (no semantics) — 24 tests
- [x] §5 conformance vectors covered structurally against `parseQuery`
- [x] `resolveCalculatorState` / `resolvePlotState` exist, tested without a page render
- [x] `migrateUrl` seam wired into `url-version.ts`
- [x] No logic duplicated between calculator and plot modules
- [x] All existing unit/contract tests pass (behaviour preserved)
- [x] New tokenization-layer unit tests
- [x] E2E unchanged & expected green (behaviour preserved — `eunit`/`mode`/`qfocus` still parse)
- [x] v1 retired: `MIN_SUPPORTED_URL_MAJOR = 2`, version-aware "unsupported link" banner
- [x] Spec updated (v8: §3.4 retired, §3.1.5, §3.9 diagnostics, §6) + published `/docs/technical` grammar page
- [x] AI log + `CHANGELOG-AI.md`

Local: `pnpm test` (1624 passing — the only 3 failures are the pre-existing sandbox git-signing issue in `guard-forbidden-files.test.ts`, which passes on CI), `check`, `lint`, `format:check`, `build` all green.

### Out of scope (potential follow-ups)
- Modernising the canonical **wire format** to the aspirational v2 tokens (`uanchor`-always, `calc`/`runit`/`sunit`) — deliberately deferred to keep behaviour stable.
- Routing the resolver `Diagnostic[]` into a richer in-app banner beyond the version notice.

Draft until CI confirms green.

https://claude.ai/code/session_019bSdTmM6jwRV8HNeVp2Mmv

### CODE REVIEW

File: src/lib/utils/calculator-url.ts [hunk 1]
```diff
@@ -904,3 +790,16 @@ export function decodeCalculatorUrl(rawParams: URLSearchParams): CalculatorUrlSt
...
+ * Decode calculator URL state from a raw query string or `URLSearchParams`.
+ * Thin wrapper over the layered pipeline: extdata is resolved on the original
+ * params (ordered, may repeat), then `parseQuery` tokenizes and
+ * `resolveCalculatorState` applies semantics.
+ */
+export function decodeCalculatorUrl(input: URLSearchParams | string): CalculatorUrlState {
+  const params = typeof input === "string" ? new URLSearchParams(input) : input;
+  const { sources } = parseExtdataParams(params);
+  const ast = parseQuery(input);
+  return resolveCalculatorState(ast, sources);
```
[the remark this answers is not in the document]
Comment (grzanka) [reply src/lib/utils/calculator-url.ts]: Fixed in 3023b1e. Two parts: (1) the grammar is now total — an empty-key/leading-`=` segment is tokenized as an unknown pair rather than failing, so a malformed bit like `?=oops` no longer throws *and* doesn't discard the valid params around it (`particle=1&=oops` still yields `particle=1`); (2) `decodeCalculatorUrl` defensively catches `UrlParseError` and falls back to defaults. Regression test added in `url-resolve.test.ts` ("recovers (no throw) from a malformed query, keeping valid params").

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/lib/utils/plot-url.ts [hunk 2]
```diff
@@ -481,3 +382,15 @@ export function decodePlotUrl(rawParams: URLSearchParams): PlotUrlDecoded {
...
+/**
+ * Decode plot URL state from a raw query string or `URLSearchParams`. Thin
+ * wrapper: extdata is resolved on the original params (ordered, may repeat),
+ * then `parseQuery` tokenizes and `resolvePlotState` applies semantics.
+ */
+export function decodePlotUrl(input: URLSearchParams | string): PlotUrlDecoded {
+  const params = typeof input === "string" ? new URLSearchParams(input) : input;
+  const { sources } = parseExtdataParams(params);
+  const ast = parseQuery(input);
+  return resolvePlotState(ast, sources);
```
[the remark this answers is not in the document]
Comment (grzanka) [reply src/lib/utils/plot-url.ts]: Fixed in 3023b1e — same fix as the calculator side: grammar totality plus a defensive `UrlParseError` catch in `decodePlotUrl` that falls back to defaults, so the plot restore path can't crash on a malformed query. Covered by the new `decodePlotUrl(new URLSearchParams("=oops"))` no-throw regression test.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/lib/utils/url-grammar.peggy [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+  = ids:EntityIdList EndOfValue { return { ids, raw: text() }; }
+  / raw:RawValue { return { ids: null, raw }; }
+EntityIdList
+  = head:EntityId tail:("," @EntityId)* { return [head, ...tail]; }
+
+// ── known scalar params (value kept verbatim) ────────────────────────────────
+ScalarPair
+  = key:ScalarKey "=" value:RawValue {
+      return { type: "scalar", key, value, span: spanOf(location()) };
+    }
```
[the remark this answers is not in the document]
Comment (grzanka) [reply src/lib/utils/url-grammar.peggy:130]: Fixed in 3023b1e. `decode*` now always normalize the input through `URLSearchParams` before tokenizing, and `URLSearchParams` decodes the key (`%70article` → `particle`) before `toString()` hands it to `parseQuery` — so the tokenizer classifies it as the `particle` scalar exactly as the legacy decoder did. Regression test added: `decodeCalculatorUrl(new URLSearchParams("%70article=6")).particleId === 6` (and the raw-string overload too).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/lib/utils/url-version.ts [hunk 4]
```diff
@@ -16,3 +23,20 @@ export function negotiateVersion(version: string | null | undefined): VersionNeg
   if (v === CURRENT_URL_MAJOR) return { status: "ok" };
   return { status: "mismatch", version: v };
```
[the remark this answers is not in the document]
Comment (grzanka) [reply src/lib/utils/url-version.ts:24]: Good catch on the contract nuance. Worth clarifying what already happens vs. what's proposed:

**The calculation _is_ halted on a version mismatch today.** The orchestrator gates it on `urlVersionMismatch` — `calculator-page-orchestrator.svelte.ts:247` (`if (this.urlVersionMismatch !== null) return;` before `triggerCalculation()`), and the multi-program / multi-entity / inverse calc setups all receive `urlVersionMismatch` and gate the same way. So an unsupported `urlv=1`/future link produces **no results** — only the banner + "Load defaults" (the E2E `urlv=999` scenario asserts the STP cell stays empty). That matches the spec's "halt the URL-driven calc".

What still happens is that entity selection / inputs are *hydrated* from the link (so the user sees what the shared link pointed at) even though nothing is computed. Fully suppressing that hydration means gating `appInit.initialize(...)` / `decodeCalculatorUrl(...)` on the negotiation result and re-validating the Load-defaults recovery path — a change to the (pre-existing) page-init flow on both the calculator and plot orchestrators, outside the parser-refactor core of this PR.

I've left this thread open for @grzanka to decide: keep the current calc-halt (and I'll tighten the spec wording so it doesn't over-promise), or I implement the stricter input-level gating as a follow-up. Happy to do either.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (grzanka) [reply src/lib/utils/url-version.ts:24]: Implemented the stricter input-level gating (option b) in 67e1d14.

An unsupported `urlv` now hydrates **nothing** — entity selection, energy rows, advanced mode, and plot series all stay at defaults, not just the calc being halted:

- **Calculator orchestrator**: version negotiation reordered ahead of advanced-mode init; `initAdvancedModeFromUrl` / `appInit.initialize` / `decodeCalculatorUrl` / `decodeMultiProgramUrl` all receive empty params when mismatched. (Multi-program is gated too because advanced mode can be active via the persisted preference even when the link's `mode` is suppressed, and it reads `window.location.search` directly.)
- **Plot orchestrator**: moved version negotiation to a synchronous pre-effect step so it resolves before init/restore fire (the old free-standing effect had no ordering guarantee); gated init + advanced mode; `setupPlotUrlRestore` now takes a mismatch getter and skips restoration while still flipping `urlInitialized` so URL-sync emits canonical defaults.
- **Load-defaults recovery re-validated**: since init already produces defaults, clearing the mismatch just opens the calc gate → computes from defaults. E2E 1b (`stp-cell-0 > 0` after Load defaults) holds because the default energy row is `"100"` with the default proton/water selection; 1a (empty cell + banner) also holds.
- Spec §3.1 steps 5c–5e now state explicitly that no state is hydrated, and the stale step-15 migration-banner line was removed.

`pnpm check` / `lint` / `format:check` green; unit suite green apart from the pre-existing sandbox git-signing failures in `guard-forbidden-files.test.ts`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by grzanka** [comment]
@copilot resolve the merge conflicts in this pull request

**Comment by grzanka** [comment]
Resolved the merge conflict with `master` (09ced9d). The only conflict was in `CHANGELOG-AI.md` — both sides prepended entries — so I kept all of them in date order (`docs/ai-logs/README.md` auto-merged cleanly). After merging master's #660/#646/#629 work: `pnpm check` / `format:check` green, unit suite 1642 passing (only the pre-existing sandbox git-signing failures in `guard-forbidden-files.test.ts` remain, green on CI).

---
_Generated by [Claude Code](https://claude.ai/code)_
